qiskit/optimization/algorithms/optimization_algorithm.py

```diff
@@ -104,24 +104,51 @@ class OptimizationResult:
         results: The original results object returned from the optimization algorithm. This can
             contain more information than only the optimal value and function value.
         status: The termination status of the algorithm.
+        variables: The list of variables under optimization.
     """
 
     Status = OptimizationResultStatus
 
     def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
                  results: Optional[Any] = None,
-                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS) -> None:
-        self._val = x
-        self._fval = fval
-        self._results = results
-        self._status = status
+                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS,
+                 variables: Optional[List[Variable]] = None) -> None:
+        self.x = x
+        self.variables = variables
+        self.fval = fval
+        self.results = results
+        self.status = status
 
     def __repr__(self):
-        return '([{}] / {} / {})'.format(','.join([str(x_) for x_ in self.x]), self.fval,
-                                         self.status)
+        return 'optimal variables: [{}]\noptimal function value: {}\nstatus: {}'\
+            .format(','.join([str(x_) for x_ in self.x]), self.fval, self.status.name)
+
+    def __getitem__(self, item: Union[int, str]):
+        if isinstance(item, int):
+            return self.x[item]
+        if isinstance(item, str):
+            return self.var_dict[item]
+        raise QiskitOptimizationError("Integer or string parameter required.")
```

Comment:
It would be nice to display `item` in the error message. It will be more informative.